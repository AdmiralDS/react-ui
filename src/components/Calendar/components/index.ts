export * from './Years';
export * from './Button';
export * from './Day';
export * from './DayNames';
export * from './Month';
export * from './Months';
export * from './Panel';
export * from './Week';
