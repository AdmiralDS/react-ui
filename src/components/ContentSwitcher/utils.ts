type SwitcherFocusCalculator = (parent: HTMLDivElement | null, child: Element | null) => Element | null;

export const moveFocus = (
  parent: HTMLDivElement | null,
  currentFocus: Element | null,
  calcNextFocus: SwitcherFocusCalculator,
): void => {
  let wrappedOnce = false;
  let nextFocus = calcNextFocus(parent, currentFocus);
  while (nextFocus) {
    if (nextFocus === parent?.firstChild) {
      if (wrappedOnce) {
        return;
      }
      wrappedOnce = true;
    }

    if (nextFocus.hasAttribute('disabled')) {
      nextFocus = calcNextFocus(parent, nextFocus);
    } else {
      nextFocus.focus();
      return;
    }
  }
};

export const nextItem = (parent: Element | null, child: Element | null): Element | null => {
  if (child?.nextElementSibling) {
    return child.nextElementSibling;
  }
  return parent?.firstElementChild ?? null;
};

export const previousItem = (parent: Element | null, child: Element | null): Element | null => {
  if (child?.previousElementSibling) {
    return child.previousElementSibling;
  }
  return parent?.lastElementChild ?? null;
};
