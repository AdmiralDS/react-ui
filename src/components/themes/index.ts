export * from './light';
export * from './dark';
export type { Theme } from './common';
