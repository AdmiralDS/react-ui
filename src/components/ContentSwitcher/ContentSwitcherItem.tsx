import { forwardRef } from 'react';
import type { ButtonHTMLAttributes } from 'react';
import styled, { css } from 'styled-components';

import { typography } from '#src/components/Typography';
import { BadgeComponent } from '#src/components/Badge';
import type { Adaptive } from './ContentSwitcherComponent';

export interface ContentSwitcherItemProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  /** Активная кнопка */
  active?: boolean;
  /** Позволяет добавлять  миксин созданный с помощью styled css  */
  cssMixin?: ReturnType<typeof css>;
  adaptive?: Adaptive;
}

const colorMixin = css<{ $active?: boolean; disabled?: boolean }>`
  color: ${({ $active, disabled, theme }) => {
    if ($active) {
      return `var(--admiral-color-Neutral_Neutral90, ${theme.color['Neutral/Neutral 90']})`;
    }
    if (disabled) {
      return `var(--admiral-color-Neutral_Neutral30, ${theme.color['Neutral/Neutral 30']})`;
    }
    return `var(--admiral-color-Neutral_Neutral50, ${theme.color['Neutral/Neutral 50']})`;
  }};
  [data-appearance~='primary'] && {
    color: ${({ $active, disabled, theme }) => {
      if ($active) {
        return `var(--admiral-color-Special_StaticWhite, ${theme.color['Special/Static White']})`;
      }
      if (disabled) {
        return `var(--admiral-color-Neutral_Neutral30, ${theme.color['Neutral/Neutral 30']})`;
      }
      return `var(--admiral-color-Primary_Primary60Main, ${theme.color['Primary/Primary 60 Main']})`;
    }};
  }
`;

const adaptiveMixin = css`
  width: 100%;
  display: flex;
  justify-content: center;
`;

const BORDER_RADIUS = 22;

export const ContentSwitcherItemButton = styled.button<{
  $active?: boolean;
  $adaptive?: Adaptive;
  $cssMixin?: ReturnType<typeof css>;
}>`
  border: none;
  border-radius: ${BORDER_RADIUS}px;
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
  display: flex;
  align-items: center;
  position: relative;
  outline: 0;
  background-color: ${({ theme, disabled, $active }) =>
    $active && !disabled
      ? `var(--admiral-color-Neutral_Neutral00, ${theme.color['Neutral/Neutral 00']})`
      : 'transparent'};

  ${typography['Button/Button 1']}
  white-space: nowrap;

  ${colorMixin};
  ${({ $adaptive }) => ($adaptive === 'fill' ? adaptiveMixin : '')}

  &:hover:not(:disabled) {
    ${({ $active, disabled, theme }) =>
      !$active && !disabled ? `background: var(--admiral-color-Opacity_Hover, ${theme.color['Opacity/Hover']});` : ''};
  }
  &:active:not(:disabled) {
    ${({ $active, disabled, theme }) =>
      !$active && !disabled ? `background: var(--admiral-color-Opacity_Press, ${theme.color['Opacity/Press']});` : ''};
  }

  & *[fill^='#'] {
    fill: ${({ disabled, theme }) => {
      if (disabled) {
        return `var(--admiral-color-Neutral_Neutral30, ${theme.color['Neutral/Neutral 30']})`;
      }
      return `var(--admiral-color-Neutral_Neutral50, ${theme.color['Neutral/Neutral 50']})`;
    }};
  }

  [data-appearance~='primary'] && *[fill^='#'] {
    fill: ${({ $active, disabled, theme }) => {
      if ($active) {
        return `var(--admiral-color-Special_StaticWhite, ${theme.color['Special/Static White']})`;
      }
      if (disabled) {
        return `var(--admiral-color-Neutral_Neutral30, ${theme.color['Neutral/Neutral 30']})`;
      }
      return `var(--admiral-color-Primary_Primary60Main, ${theme.color['Primary/Primary 60 Main']})`;
    }};
  }

  & ${BadgeComponent} {
    background-color: ${(p) =>
      p.$active
        ? `var(--admiral-color-Neutral_Neutral50, ${p.theme.color['Neutral/Neutral 50']})`
        : `var(--admiral-color-Neutral_Neutral00, ${p.theme.color['Neutral/Neutral 00']})`};
    color: ${({ theme, disabled, $active }) => {
      if ($active) {
        return `var(--admiral-color-Special_StaticWhite, ${theme.color['Special/Static White']})`;
      }
      if (disabled) {
        return `var(--admiral-color-Neutral_Neutral30, ${theme.color['Neutral/Neutral 30']})`;
      }
      return `var(--admiral-color-Neutral_Neutral50, ${theme.color['Neutral/Neutral 50']})`;
    }};
  }

  [data-appearance~='primary'] && ${BadgeComponent} {
    background-color: ${({ $active, disabled, theme }) => {
      if ($active) {
        return `var(--admiral-color-Special_StaticWhite, ${theme.color['Special/Static White']})`;
      }
      if (disabled) {
        return `var(--admiral-color-Neutral_Neutral10, ${theme.color['Neutral/Neutral 10']})`;
      }
      return `var(--admiral-color-Primary_Primary60Main, ${theme.color['Primary/Primary 60 Main']})`;
    }};
    color: ${({ theme, disabled, $active }) => {
      if ($active) {
        return `var(--admiral-color-Primary_Primary60Main, ${theme.color['Primary/Primary 60 Main']})`;
      }
      if (disabled) {
        return `var(--admiral-color-Neutral_Neutral30, ${theme.color['Neutral/Neutral 30']})`;
      }
      return `var(--admiral-color-Special_StaticWhite, ${theme.color['Special/Static White']})`;
    }};
  }

  [data-appearance~='primary'] && {
    background-color: ${({ theme, disabled, $active }) =>
      $active && !disabled
        ? `var(--admiral-color-Primary_Primary60Main, ${theme.color['Primary/Primary 60 Main']})`
        : 'transparent'};
    &:hover:not(:disabled) {
      ${({ $active, disabled, theme }) =>
        !$active && !disabled
          ? `background: var(--admiral-color-Opacity_Hover, ${theme.color['Opacity/Hover']});`
          : ''};
    }
    &:active:not(:disabled) {
      ${({ $active, disabled, theme }) =>
        !$active && !disabled
          ? `background: var(--admiral-color-Opacity_Press, ${theme.color['Opacity/Press']});`
          : ''};
    }
  }

  ${(p) => p.$cssMixin && p.$cssMixin}
`;

export const ContentSwitcherItem = forwardRef<HTMLButtonElement, ContentSwitcherItemProps>(
  ({ active, cssMixin, type = 'button', adaptive, ...props }: ContentSwitcherItemProps, ref) => {
    return (
      <ContentSwitcherItemButton
        ref={ref}
        {...{ type, ...props }}
        $adaptive={adaptive}
        $active={active}
        $cssMixin={cssMixin}
      />
    );
  },
);
