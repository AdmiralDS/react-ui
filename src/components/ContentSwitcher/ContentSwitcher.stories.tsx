import React, { useState } from 'react';
import styled from 'styled-components';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import { withDesign } from 'storybook-addon-designs';
import { ContentSwitcher } from '#src/components/ContentSwitcher';
import { ContentSwitcherItem } from '#src/components/ContentSwitcher/ContentSwitcherItem';
import { ReactComponent as StarSolid } from '@admiral-ds/icons/build/system/StarSolid.svg';
import { ReactComponent as DeleteOutline } from '@admiral-ds/icons/build/system/DeleteOutline.svg';
import { ReactComponent as EmailOutline } from '@admiral-ds/icons/build/system/EmailOutline.svg';
import { ReactComponent as HomeOutline } from '@admiral-ds/icons/build/system/HomeOutline.svg';
import { ReactComponent as PrintOutline } from '@admiral-ds/icons/build/system/PrintOutline.svg';
import { T } from '#src/components/T';
import { Badge } from '#src/components/Badge';

const Divider = styled.div`
  width: 10px;
`;

const Separator = styled.div`
  height: 20px;
`;

export default {
  title: 'Admiral-2.1/ContentSwitcher',
  decorators: [withDesign],
  component: ContentSwitcher,
  parameters: {
    docs: {
      source: {
        code: null,
      },
    },
    design: [
      {
        type: 'figma',
        url: 'https://www.figma.com/file/EGEGZsx8WhdxpmFKu8J41G/Admiral-2.1-UI-Kit?node-id=7582%3A70132',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/EGEGZsx8WhdxpmFKu8J41G/Admiral-2.1-UI-Kit?node-id=7582%3A70157',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/EGEGZsx8WhdxpmFKu8J41G/Admiral-2.1-UI-Kit?node-id=7582%3A70180',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/EGEGZsx8WhdxpmFKu8J41G/Admiral-2.1-UI-Kit?node-id=7582%3A70203',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/EGEGZsx8WhdxpmFKu8J41G/Admiral-2.1-UI-Kit?node-id=7582%3A70134',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/EGEGZsx8WhdxpmFKu8J41G/Admiral-2.1-UI-Kit?node-id=7582%3A70226',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/EGEGZsx8WhdxpmFKu8J41G/Admiral-2.1-UI-Kit?node-id=7582%3A70249',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/EGEGZsx8WhdxpmFKu8J41G/Admiral-2.1-UI-Kit?node-id=7582%3A70273',
      },
    ],
  },
  argTypes: {
    dimension: {
      control: {
        type: 'radio',
        options: ['l', 'm', 's'],
      },
    },
  },
} as ComponentMeta<typeof ContentSwitcher>;

const MobileWrapper = styled.div`
  width: 320px;
  height: 100px;
  padding-top: 16px;
  box-sizing: border-box;
  border: ${(p) => `8px solid ${p.theme.color['Neutral/Neutral 50']}`};
  border-bottom-style: none;
  background-color: ${(p) => p.theme.color['Neutral/Neutral 00']};
  border-top-left-radius: 24px;
  border-top-right-radius: 24px;
`;

const SwitcherWrapper = styled.div`
  overflow-x: auto;
  width: 100%;
  padding: 0 8px;
  box-sizing: border-box;
  -ms-overflow-style: none; /* IE and Edge */
  scrollbar-width: none; /* Firefox */
  &::-webkit-scrollbar {
    display: none;
  }
`;

const Wrapper = styled.div`
  display: flex;
  align-items: center;
`;

const ContentSwitcherDemo: ComponentStory<typeof ContentSwitcher> = () => {
  const list = [
    { title: 'Active button', content: '1' },
    { title: 'Default button', content: '2' },
    { title: 'Disabled button', content: '3', disabled: true },
  ];
  return (
    <>
      <T font="Body/Body 1 Long" as="div">
        Dimension - l
      </T>
      <Separator />
      <ContentSwitcher dimension={'l'}>
        {list.map((item, index) => {
          return (
            <ContentSwitcherItem key={index} active={index === 0} disabled={item.disabled}>
              {item.title}
            </ContentSwitcherItem>
          );
        })}
      </ContentSwitcher>
      <Separator />
      <ContentSwitcher dimension={'l'} appearance="primary">
        {list.map((item, index) => {
          return (
            <ContentSwitcherItem key={index} active={index === 0} disabled={item.disabled}>
              {item.title}
            </ContentSwitcherItem>
          );
        })}
      </ContentSwitcher>
      <Separator />
      <T font="Body/Body 1 Long" as="div">
        Dimension - m
      </T>
      <Separator />
      <ContentSwitcher dimension={'m'}>
        {list.map((item, index) => {
          return (
            <ContentSwitcherItem key={index} active={index === 0} disabled={item.disabled}>
              {item.title}
            </ContentSwitcherItem>
          );
        })}
      </ContentSwitcher>
      <Separator />
      <T font="Body/Body 1 Long" as="div">
        Dimension - s
      </T>
      <Separator />
      <ContentSwitcher dimension={'s'}>
        {list.map((item, index) => {
          return (
            <ContentSwitcherItem key={index} active={index === 0} disabled={item.disabled}>
              {item.title}
            </ContentSwitcherItem>
          );
        })}
      </ContentSwitcher>
      <Separator />
      <T font="Body/Body 1 Long" as="div">
        Mobile Adaptive
      </T>
      <MobileWrapper>
        <SwitcherWrapper>
          <ContentSwitcher dimension={'m'}>
            {list.map((item, index) => {
              return (
                <ContentSwitcherItem key={index} active={index === 0} disabled={item.disabled}>
                  {item.title}
                </ContentSwitcherItem>
              );
            })}
          </ContentSwitcher>
        </SwitcherWrapper>
      </MobileWrapper>
    </>
  );
};

const ContentSwitcherWithIconDemo: ComponentStory<typeof ContentSwitcher> = () => {
  return (
    <>
      <ContentSwitcher dimension={'l'}>
        <ContentSwitcherItem active>
          <StarSolid width={20} />
          <Divider />
          Active button
        </ContentSwitcherItem>
        <ContentSwitcherItem>
          <StarSolid width={20} />
          <Divider />
          Default button
        </ContentSwitcherItem>
        <ContentSwitcherItem disabled>
          Disabled button
          <Divider />
          <StarSolid width={17} />
        </ContentSwitcherItem>
      </ContentSwitcher>
    </>
  );
};

const ContentSwitcherIconOnlyDemo: ComponentStory<typeof ContentSwitcher> = () => {
  const [activeLPrimary, setActiveLPrimary] = useState(0);
  const [activeMPrimary, setActiveMPrimary] = useState(0);
  const [activeSPrimary, setActiveSPrimary] = useState(0);
  const [activeLSecondary, setActiveLSecondary] = useState(0);
  const [activeMSecondary, setActiveMSecondary] = useState(0);
  const [activeSSecondary, setActiveSSecondary] = useState(0);

  const list = [
    {
      title: <DeleteOutline width={20} />,
    },
    {
      title: <EmailOutline width={20} />,
    },
    {
      title: <HomeOutline width={20} />,
    },
    {
      title: <PrintOutline width={20} />,
    },
  ];

  return (
    <>
      <Wrapper>
        <ContentSwitcher dimension={'l'} appearance="primary">
          {list.map((item, index) => (
            <ContentSwitcherItem key={index} active={index === activeLPrimary} onClick={() => setActiveLPrimary(index)}>
              {item.title}
            </ContentSwitcherItem>
          ))}
        </ContentSwitcher>
        <Divider />
        <ContentSwitcher dimension={'m'} appearance="primary">
          {list.map((item, index) => (
            <ContentSwitcherItem key={index} active={index === activeMPrimary} onClick={() => setActiveMPrimary(index)}>
              {item.title}
            </ContentSwitcherItem>
          ))}
        </ContentSwitcher>
        <Divider />
        <ContentSwitcher dimension={'s'} appearance="primary">
          {list.map((item, index) => (
            <ContentSwitcherItem key={index} active={index === activeSPrimary} onClick={() => setActiveSPrimary(index)}>
              {item.title}
            </ContentSwitcherItem>
          ))}
        </ContentSwitcher>
      </Wrapper>
      <Separator />
      <Wrapper>
        <ContentSwitcher dimension={'l'} appearance="secondary">
          {list.map((item, index) => (
            <ContentSwitcherItem
              key={index}
              active={index === activeLSecondary}
              onClick={() => setActiveLSecondary(index)}
            >
              {item.title}
            </ContentSwitcherItem>
          ))}
        </ContentSwitcher>
        <Divider />
        <ContentSwitcher dimension={'m'} appearance="secondary">
          {list.map((item, index) => (
            <ContentSwitcherItem
              key={index}
              active={index === activeMSecondary}
              onClick={() => setActiveMSecondary(index)}
            >
              {item.title}
            </ContentSwitcherItem>
          ))}
        </ContentSwitcher>
        <Divider />
        <ContentSwitcher dimension={'s'} appearance="secondary">
          {list.map((item, index) => (
            <ContentSwitcherItem
              key={index}
              active={index === activeSSecondary}
              onClick={() => setActiveSSecondary(index)}
            >
              {item.title}
            </ContentSwitcherItem>
          ))}
        </ContentSwitcher>
      </Wrapper>
    </>
  );
};

const ContentSwitcherWithBadgeDemo: ComponentStory<typeof ContentSwitcher> = () => {
  const [active, setActive] = useState(0);

  const list = [
    {
      title: (
        <>
          Active button <Badge style={{ marginLeft: 8 }}>3</Badge>
        </>
      ),
    },
    {
      title: (
        <>
          Default button <Badge style={{ marginLeft: 8 }}>6</Badge>
        </>
      ),
    },
    {
      title: (
        <>
          Disabled button <Badge style={{ marginLeft: 8 }}>6</Badge>
        </>
      ),
      disabled: true,
    },
  ];

  return (
    <>
      <ContentSwitcher dimension={'l'} appearance="primary">
        {list.map((item, index) => (
          <ContentSwitcherItem
            key={index}
            active={index === active}
            onClick={() => setActive(index)}
            disabled={item.disabled}
          >
            {item.title}
          </ContentSwitcherItem>
        ))}
      </ContentSwitcher>
    </>
  );
};

const Template: ComponentStory<typeof ContentSwitcher> = ({ dimension, className, ...props }) => {
  const [active, setActive] = useState(0);

  const list = [
    {
      title: (
        <>
          <StarSolid width={17} style={{ marginRight: 8 }} />
          Active button
        </>
      ),
    },
    {
      title: (
        <>
          <StarSolid width={17} style={{ marginRight: 8 }} />
          Default button <Badge style={{ marginLeft: 8 }}>6</Badge>
        </>
      ),
    },
    { title: 'Disabled button', disabled: true },
  ];

  return (
    <>
      <ContentSwitcher dimension={dimension} className={className} {...props}>
        {list.map((item, index) => (
          <ContentSwitcherItem
            key={index}
            active={index === active}
            onClick={() => setActive(index)}
            disabled={item.disabled}
          >
            {item.title}
          </ContentSwitcherItem>
        ))}
      </ContentSwitcher>
    </>
  );
};

export const Example = ContentSwitcherDemo.bind({});
export const ContentSwitcherWithIcon = ContentSwitcherWithIconDemo.bind({});
export const ContentSwitcherIconOnly = ContentSwitcherIconOnlyDemo.bind({});
export const ContentSwitcherWithBadge = ContentSwitcherWithBadgeDemo.bind({});
export const Playground = Template.bind({});
Example.storyName = 'Базовый пример';
ContentSwitcherWithIcon.storyName = 'ContentSwitcher с иконками';
ContentSwitcherIconOnly.storyName = 'ContentSwitcher только с иконками';
ContentSwitcherWithBadge.storyName = 'ContentSwitcher с Badge';
Playground.storyName = 'ContentSwitcher. Playground';
Playground.args = {};
