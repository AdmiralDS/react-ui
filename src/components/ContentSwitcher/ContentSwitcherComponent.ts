import type { HTMLAttributes } from 'react';
import styled, { css } from 'styled-components';
import { ContentSwitcherItemButton } from '#src/components/ContentSwitcher/ContentSwitcherItem';

export type Dimension = 'l' | 'm' | 's';
export type Adaptive = 'fill';

export const SIZE = {
  l: 40,
  m: 32,
  s: 24,
};

type ContentSwitcherComponentProps = HTMLAttributes<HTMLDivElement> & {
  $dimension: Dimension;
  $adaptive?: Adaptive;
};

const BORDER_RADIUS = {
  l: 24,
  m: 20,
  s: 16,
};

const padding = {
  l: 20,
  m: 16,
  s: 12,
};

const adaptiveMixin = css`
  width: 100%;
  overflow: hidden;
  min-width: fit-content;
`;

export const ContentSwitcherComponent = styled.div<ContentSwitcherComponentProps>`
  background: var(--admiral-color-Opacity_Neutral8, ${(p) => p.theme.color['Opacity/Neutral 8']});
  border: 2px solid transparent;
  border-radius: ${({ $dimension }) => BORDER_RADIUS[$dimension]}px;
  display: inline-flex;
  height: ${({ $dimension }) => SIZE[$dimension]}px;
  padding: 2px;

  ${({ $adaptive }) => ($adaptive === 'fill' ? adaptiveMixin : '')}

  &[data-appearance~='primary'] {
    border: 2px solid var(--admiral-color-Primary_Primary60Main, ${(p) => p.theme.color['Primary/Primary 60 Main']});
    background-color: transparent;
  }

  ${ContentSwitcherItemButton} {
    height: 100%;
    padding: ${({ $dimension }) => `0 ${padding[$dimension]}px`};
    font-size: ${({ $dimension }) => ($dimension === 'l' ? 16 : 14)}px;
    line-height: ${({ $dimension }) => ($dimension === 'l' ? 24 : 20)}px;
  }

  > *:not(:first-child) {
    margin-left: 8px;
  }
`;
