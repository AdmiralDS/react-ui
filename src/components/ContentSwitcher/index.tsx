import type { HTMLAttributes, KeyboardEvent, ReactNode } from 'react';
import { useRef } from 'react';
import type { Adaptive, Dimension } from '#src/components/ContentSwitcher/ContentSwitcherComponent';
import { ContentSwitcherComponent } from '#src/components/ContentSwitcher/ContentSwitcherComponent';
import { moveFocus, nextItem, previousItem } from '#src/components/ContentSwitcher/utils';
import { keyboardKey } from '../common/keyboardKey';

export interface ContentSwitcherProps extends HTMLAttributes<HTMLDivElement> {
  /** Размер компонента */
  dimension?: Dimension;
  /** Элементы содержимого */
  children?: ReactNode;
  /** Стиль отображения компонента */
  appearance?: 'primary' | 'secondary';
  adaptive?: Adaptive;
}

export const ContentSwitcher = ({
  dimension = 'l',
  tabIndex = -1,
  appearance = 'secondary',
  adaptive,
  ...props
}: ContentSwitcherProps) => {
  const localRef = useRef<HTMLDivElement>(null);

  const getFocusedOption = () => {
    return ((localRef.current && localRef.current.ownerDocument) || document).activeElement;
  };

  const handleKeyDown = (e: KeyboardEvent<HTMLDivElement>) => {
    const focusedOption = getFocusedOption();
    const code = keyboardKey.getCode(e);
    if (code === keyboardKey.ArrowRight) {
      moveFocus(localRef.current, focusedOption, nextItem);
    } else if (code === keyboardKey.ArrowLeft) {
      moveFocus(localRef.current, focusedOption, previousItem);
    }
    props?.onKeyDown?.(e);
  };

  return (
    <ContentSwitcherComponent
      ref={localRef}
      onKeyDown={handleKeyDown}
      $dimension={dimension}
      tabIndex={tabIndex}
      data-appearance={appearance}
      $adaptive={adaptive}
      {...props}
    />
  );
};

ContentSwitcher.displayName = 'ContentSwitcher';
