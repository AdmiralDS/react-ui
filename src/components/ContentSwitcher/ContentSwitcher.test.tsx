import * as React from 'react';
import { ThemeProvider } from 'styled-components';
import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { ContentSwitcherItem } from '#src/components/ContentSwitcher/ContentSwitcherItem';
import { ContentSwitcher } from '#src/components/ContentSwitcher';
import { LIGHT_THEME } from '#src/components/themes';

const BUTTON_TEXT = 'Button';
const BUTTON_CLICK_TEXT = 'Button Click';
const BUTTON_DISABLED_TEXT = 'Button Disabled';

describe('ContentSwitcher', () => {
  it('should render component with Button', () => {
    const { container } = render(
      <ThemeProvider theme={LIGHT_THEME}>
        <ContentSwitcher dimension={'m'}>
          <ContentSwitcherItem>{BUTTON_TEXT}</ContentSwitcherItem>
        </ContentSwitcher>
      </ThemeProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render component with Buttons', () => {
    const { container } = render(
      <ThemeProvider theme={LIGHT_THEME}>
        <ContentSwitcher dimension={'m'}>
          <ContentSwitcherItem>{BUTTON_TEXT}</ContentSwitcherItem>
          <ContentSwitcherItem>{BUTTON_TEXT}</ContentSwitcherItem>
          <ContentSwitcherItem>{BUTTON_TEXT}</ContentSwitcherItem>
          <ContentSwitcherItem>{BUTTON_TEXT}</ContentSwitcherItem>
        </ContentSwitcher>
      </ThemeProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test('renders the ContentSwitcher', () => {
    render(
      <ThemeProvider theme={LIGHT_THEME}>
        <ContentSwitcher dimension={'m'}>
          <ContentSwitcherItem>{BUTTON_TEXT}</ContentSwitcherItem>
        </ContentSwitcher>
      </ThemeProvider>,
    );
    expect(screen.getByRole('button')).toHaveTextContent(`${BUTTON_TEXT}`);
  });

  test('should call onClick when user clicks on last button', () => {
    const handleClick = jest.fn();
    render(
      <ThemeProvider theme={LIGHT_THEME}>
        <ContentSwitcher dimension={'m'}>
          <ContentSwitcherItem onClick={handleClick}>{BUTTON_TEXT}</ContentSwitcherItem>
          <ContentSwitcherItem onClick={handleClick}>{BUTTON_CLICK_TEXT}</ContentSwitcherItem>
        </ContentSwitcher>
      </ThemeProvider>,
    );

    fireEvent.click(screen.getByText(`${BUTTON_CLICK_TEXT}`));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  test('Dont should call onClick when user clicks on last button', () => {
    const handleClick = jest.fn();
    render(
      <ThemeProvider theme={LIGHT_THEME}>
        <ContentSwitcher dimension={'m'}>
          <ContentSwitcherItem onClick={handleClick}>{BUTTON_TEXT}</ContentSwitcherItem>
          <ContentSwitcherItem disabled onClick={handleClick}>
            {BUTTON_CLICK_TEXT}
          </ContentSwitcherItem>
        </ContentSwitcher>
      </ThemeProvider>,
    );

    fireEvent.click(screen.getByText(`${BUTTON_CLICK_TEXT}`));
    expect(handleClick).toHaveBeenCalledTimes(0);
  });

  test('should focus on component when user on press key tab', async () => {
    const user = userEvent.setup();
    const handleClick = jest.fn();
    render(
      <ThemeProvider theme={LIGHT_THEME}>
        <ContentSwitcher data-testid="test" dimension={'m'}>
          <ContentSwitcherItem data-testid={BUTTON_CLICK_TEXT} onClick={handleClick}>
            {BUTTON_CLICK_TEXT}
          </ContentSwitcherItem>
          <ContentSwitcherItem data-testid={BUTTON_TEXT} onClick={handleClick}>
            {BUTTON_TEXT}
          </ContentSwitcherItem>
          <ContentSwitcherItem disabled onClick={handleClick}>
            {BUTTON_DISABLED_TEXT}
          </ContentSwitcherItem>
        </ContentSwitcher>
      </ThemeProvider>,
    );
    expect(document.body).toHaveFocus();
    await user.tab();
    expect(screen.getByTestId(BUTTON_CLICK_TEXT)).toHaveFocus();
    await user.tab();
    expect(screen.getByTestId(BUTTON_TEXT)).toHaveFocus();
    await user.tab();
    expect(document.body).toHaveFocus();
  });
});
