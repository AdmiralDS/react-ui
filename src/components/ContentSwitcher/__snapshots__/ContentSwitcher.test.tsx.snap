// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentSwitcher should render component with Button 1`] = `
.c2 {
  border: none;
  border-radius: 22px;
  cursor: pointer;
  display: flex;
  align-items: center;
  position: relative;
  outline: 0;
  background-color: transparent;
  font-family: var(--admiral-font-family, 'VTB Group UI', sans-serif);
  font-style: normal;
  font-weight: var(--admiral-font-weight-Button_Button1, 500);
  font-size: var(--admiral-font-size-Button_Button1, 16px);
  line-height: var(--admiral-line-height-Button_Button1, 24px);
  font-feature-settings: 'tnum' on,'lnum' on;
  text-rendering: geometricPrecision;
  white-space: nowrap;
  color: var(--admiral-color-Neutral_Neutral50, #74767B);
}

[data-appearance~='primary'] .c2.c2 {
  color: var(--admiral-color-Primary_Primary60Main, #0062FF);
}

.c2:hover:not(:disabled) {
  background: var(--admiral-color-Opacity_Hover, #0000000D);
}

.c2:active:not(:disabled) {
  background: var(--admiral-color-Opacity_Press, #0000001F);
}

.c2 *[fill^='#'] {
  fill: var(--admiral-color-Neutral_Neutral50, #74767B);
}

[data-appearance~='primary'] .c2.c2 *[fill^='#'] {
  fill: var(--admiral-color-Primary_Primary60Main, #0062FF);
}

[data-appearance~='primary'] .c2.c2 {
  background-color: transparent;
}

[data-appearance~='primary'] .c2.c2:hover:not(:disabled) {
  background: var(--admiral-color-Opacity_Hover, #0000000D);
}

[data-appearance~='primary'] .c2.c2:active:not(:disabled) {
  background: var(--admiral-color-Opacity_Press, #0000001F);
}

.c0 {
  background: var(--admiral-color-Opacity_Neutral8, #00000014);
  border: 2px solid transparent;
  border-radius: 20px;
  display: inline-flex;
  height: 32px;
  padding: 2px;
}

.c0[data-appearance~='primary'] {
  border: 2px solid var(--admiral-color-Primary_Primary60Main, #0062FF);
  background-color: transparent;
}

.c0 .c1 {
  height: 100%;
  padding: 0 16px;
  font-size: 14px;
  line-height: 20px;
}

.c0 >*:not(:first-child) {
  margin-left: 8px;
}

<div>
  <div
    class="c0"
    data-appearance="secondary"
    tabindex="-1"
  >
    <button
      class="c1 c2"
      type="button"
    >
      Button
    </button>
  </div>
</div>
`;

exports[`ContentSwitcher should render component with Buttons 1`] = `
.c2 {
  border: none;
  border-radius: 22px;
  cursor: pointer;
  display: flex;
  align-items: center;
  position: relative;
  outline: 0;
  background-color: transparent;
  font-family: var(--admiral-font-family, 'VTB Group UI', sans-serif);
  font-style: normal;
  font-weight: var(--admiral-font-weight-Button_Button1, 500);
  font-size: var(--admiral-font-size-Button_Button1, 16px);
  line-height: var(--admiral-line-height-Button_Button1, 24px);
  font-feature-settings: 'tnum' on,'lnum' on;
  text-rendering: geometricPrecision;
  white-space: nowrap;
  color: var(--admiral-color-Neutral_Neutral50, #74767B);
}

[data-appearance~='primary'] .c2.c2 {
  color: var(--admiral-color-Primary_Primary60Main, #0062FF);
}

.c2:hover:not(:disabled) {
  background: var(--admiral-color-Opacity_Hover, #0000000D);
}

.c2:active:not(:disabled) {
  background: var(--admiral-color-Opacity_Press, #0000001F);
}

.c2 *[fill^='#'] {
  fill: var(--admiral-color-Neutral_Neutral50, #74767B);
}

[data-appearance~='primary'] .c2.c2 *[fill^='#'] {
  fill: var(--admiral-color-Primary_Primary60Main, #0062FF);
}

[data-appearance~='primary'] .c2.c2 {
  background-color: transparent;
}

[data-appearance~='primary'] .c2.c2:hover:not(:disabled) {
  background: var(--admiral-color-Opacity_Hover, #0000000D);
}

[data-appearance~='primary'] .c2.c2:active:not(:disabled) {
  background: var(--admiral-color-Opacity_Press, #0000001F);
}

.c0 {
  background: var(--admiral-color-Opacity_Neutral8, #00000014);
  border: 2px solid transparent;
  border-radius: 20px;
  display: inline-flex;
  height: 32px;
  padding: 2px;
}

.c0[data-appearance~='primary'] {
  border: 2px solid var(--admiral-color-Primary_Primary60Main, #0062FF);
  background-color: transparent;
}

.c0 .c1 {
  height: 100%;
  padding: 0 16px;
  font-size: 14px;
  line-height: 20px;
}

.c0 >*:not(:first-child) {
  margin-left: 8px;
}

<div>
  <div
    class="c0"
    data-appearance="secondary"
    tabindex="-1"
  >
    <button
      class="c1 c2"
      type="button"
    >
      Button
    </button>
    <button
      class="c1 c2"
      type="button"
    >
      Button
    </button>
    <button
      class="c1 c2"
      type="button"
    >
      Button
    </button>
    <button
      class="c1 c2"
      type="button"
    >
      Button
    </button>
  </div>
</div>
`;
